src/Microsoft.AspNetCore.SystemWebAdapters.CoreServices/HttpApplication/HttpApplicationOptions.cs

```diff
@@ -0,0 +1,138 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Web;
+
+namespace Microsoft.AspNetCore.SystemWebAdapters;
+
+public class HttpApplicationOptions
+{
+    private readonly ModuleCollection _modules = new();
+
+    private Type _applicationType = typeof(HttpApplication);
+
+    public Type ApplicationType
+    {
+        get => _applicationType;
+        set
+        {
+            ArgumentNullException.ThrowIfNull(value);
+
+            _modules.CheckIsReadOnly();
+
+            if (!_applicationType.IsAssignableTo(typeof(HttpApplication)))
+            {
+                throw new InvalidOperationException($"Type {value.FullName} is not a valid HttpApplication");
+            }
+
+            _applicationType = value;
+        }
+    }
+
+    public IDictionary<string, Type> Modules => _modules;
+
+    internal void MakeReadOnly() => _modules.MakeReadOnly();
+
+    /// <summary>
+    /// Gets or sets the number of <see cref="HttpApplication"/> retained for reuse. In order to support modules and appplications that may contain state,
+    /// a unique instance is required for each request. This type should be set to the average number of concurrent requests expected to be seen.
+    /// </summary>
+    public int PoolSize { get; set; } = 100;
+
+    public void RegisterModule<T>(string name)
+         where T : IHttpModule
+        => Modules.Add(name, typeof(T));
```

Comment:
To allow code similar to `System.Web.HttpApplication.RegisterModule`, could we do

```cs
    public void RegisterModule<T>(string? name = null)
         where T : IHttpModule
        => Modules.Add(name ?? MakeUniqueModuleName(typeof(T)), typeof(T));

    private static string MakeUniqueModuleName(Type moduleType) => 
        $"DynamicModule_{moduleType.AssemblyQualifiedName}_{Guid.NewGuid()}".ToString(CultureInfo.InvariantCulture);
```

Alternatively, could we just add these to the `IServiceProvider` then use `GetServices` when collecting the module factories in `HttpApplicationPooledObjectPolicy`?

Reply:
sure - I've added that